crates/extensions/c8y_http_proxy/src/actor.rs

```diff
@@ -187,31 +195,113 @@ impl C8YHttpProxyActor {
+            request_internal_id.bearer_auth(jwt_token)
+        };
+
+        let token = if let Ok(token) = self.peers.jwt.await_response(()).await? {
+            self.end_point.token = token.clone();
+            token
+        } else {
+            return Err(C8YRestError::CustomError("JWT token not available".into()));
+        };
 
-        let request_internal_id = HttpRequestBuilder::get(url_get_id);
-        let res = self.execute(request_internal_id).await?;
+        let res = self
+            .execute_to_get_interal_id(
+                token,
+                self.end_point.c8y_internal_id.clone(),
+                req_build_closure,
+            )
+            .await?;
         let res = res.error_for_status()?;
 
         let internal_id_response: InternalIdResponse = res.json().await?;
         let internal_id = internal_id_response.id();
         Ok(internal_id)
     }
 
-    async fn execute(
+    async fn execute_to_get_interal_id(
         &mut self,
-        request_builder: HttpRequestBuilder,
+        token: String,
+        internal_id: String,
+        req_builder_closure: impl Fn(String, String) -> HttpRequestBuilder,
     ) -> Result<HttpResult, C8YRestError> {
         // Get a JWT token to authenticate the device
-        let request_builder = if let Ok(token) = self.peers.jwt.await_response(()).await? {
-            request_builder.bearer_auth(token)
-        } else {
-            return Err(C8YRestError::CustomError("JWT token not available".into()));
-        };
+        let request_builder = req_builder_closure(token.clone(), internal_id.clone());
 
-        // TODO Add timeout
-        // TODO Manage 403 errors
         let request = request_builder.build()?;
-        Ok(self.peers.http.await_response(request).await?)
+        let resp = self.peers.http.await_response(request).await?;
+
+        match resp {
+            Ok(response) => match response.status() {
+                StatusCode::OK => Ok(Ok(response)),
+                StatusCode::UNAUTHORIZED | StatusCode::FORBIDDEN => {
+                    let request_builder = if let Ok(token) =
+                        self.peers.jwt.await_response(()).await?
+                    {
+                        req_builder_closure(token, internal_id.clone())
+                    } else {
+                        return Err(C8YRestError::CustomError("JWT token not available".into()));
+                    };
+                    let request = request_builder.build()?;
+                    Ok(self.peers.http.await_response(request).await?)
+                }
+
+                code => Err(C8YRestError::FromHttpError(
+                    tedge_http_ext::HttpError::HttpStatusError(code),
+                )),
+            },
+
+            Err(e) => Err(C8YRestError::FromHttpError(e)),
+        }
+    }
+
+    async fn execute_retry(
+        &mut self,
+        token: String,
+        internal_id: String,
```

Comment:
I would pass neither a token nor an internal id to this method.

But use a cache maintained by `self` or `self.end_point`.

- First attempt: use the cached values
- Second attempt: clear the cache and renew  the token before retrying.

Reply:
done